src/common/state_machine.py

```diff
@@ -0,0 +1,58 @@
+from abc import ABC, abstractmethod
+
+
+class Transition:
+    def __init__(self, from_state, to_state, condition_func):
+        self.from_state = from_state
+        self.to_state = to_state
+        self.condition_func = condition_func
+
+
+class StateMachine(ABC):
+    def __init__(self):
+        self._state = None
+        self._previous_state = None
+        self._states = {}
+        self._transitions = {}
+
+    @abstractmethod
+    def _state_logic(self):
+        pass
+
+    def _get_transition(self):
+        for transition in self._transitions[self._state]:
+            if transition.condition_func():
+                return transition.to_state
+        return None
```

Comment:
No need for this line. When no return is found, Python returns `None`.

Reply:
If you ask me, I rather explicit returns :)

Reply:
Yeah, same here, I'd rather have explicit return here so it's clear